azurelinuxagent/ga/update.py

```diff
@@ -912,40 +921,100 @@ def _check_and_download_agent_if_upgrade_available(self, protocol, base_version=
+                # If there's no new incarnation, don't proceed with anything
+                requested_version = manifests[0].requested_version
+                msg = "Found requested version in manifest: {0} for incarnation: {1}".format(
+                    requested_version, incarnation)
+                logger.info(msg)
+                add_event(AGENT_NAME, op=WALAEventOperation.AgentUpgrade, is_success=True, message=msg)
+            else:
+                # If incarnation didn't change, don't process anything.
+                return False
+        else:
+            # If no requested version specified in the Goal State, follow the old auto-update logic
+            # Note: If the first Goal State contains a requested version, this timer won't start (i.e. self.last_attempt_time won't be updated).
+            # If any subsequent goal state does not contain requested version, this timer will start then, and we will
+            # download all versions available in PIR and auto-update to the highest available version on that goal state.
+            now = time.time()
+            if self.last_attempt_time is not None:
+                next_attempt_time = self.last_attempt_time + conf.get_autoupdate_frequency()
+            else:
+                next_attempt_time = now
+            if next_attempt_time > now:
+                return False
 
-            # Set the agents to those available for download at least as
-            # current as the existing agent and remove from disk any agent
-            # no longer reported to the VM.
-            # Note:
-            #  The code leaves on disk available, but blacklisted, agents
-            #  so as to preserve the state. Otherwise, those agents could be
-            #  again downloaded and inappropriately retried.
+            logger.info("[Old Auto-update logic] Checking for agent updates (family: {0})", family)
+
+            self.last_attempt_time = now
+
+        try:
+            # If we make it to this point, then either there is a requested version in a new GS (new auto-update model),
+            # or the 1hr time limit has elapsed for us to check the agent manifest for updates (old auto-update model).
+            pkg_list = ExtHandlerPackageList()
+
+            # If the requested version is the current version, don't download anything and delete all other agents from disk
+            # In this case, no need to even fetch the GA family manifest as we don't need to download any agent.
+            if requested_version is not None and requested_version == CURRENT_VERSION:
+                packages_to_download = []
+                logger.info("The requested version is running as the current version: {0}".format(requested_version))
+            else:
+                pkg_list = protocol.get_vmagent_pkgs(manifests[0])
+                packages_to_download = pkg_list.versions
```

Comment:
have you looked more at the future of the current model (downloading all available and blacklisting). I would suggest to consider if we want (as part of this change) limit the downloads to the lastest, or the lastest 2 or 3, instead of everything in storage)

Reply:
I honestly didn't think of that, not for this PR atleast. I can take that up, but that should be a separate PR with a separate sets of changes. 
Also since we're speaking of it, is deleting from PIR not good enough? Should we also limit it from the agent end?

Reply:
We tend to forget cleaning up all agents, or do it too late, or run into issues when cleaning up, and in the meanwhile millions of Vms are downloading all agents in PIR. There have been a couple of github issues in this area (I think one of them is still open).

This makes me thing we should limit the number of downloads here.

Reply:
I understand the motivation and I agree to it. That being said, I feel this should be handled in a separate PR since it would require its own testing and code refactoring. If you want I can open a task for it to keep track of it.

Reply:
yes, can be separate, thanks